### PR TITLE
Add intial api

### PR DESCRIPTION
**Pull request by calebmer**
Ok, so here are my initial ideas on the “virtualization” of the database api.
Pinging @svmnotn 😊

Here are some of the things which should probably happen pre-merge. There are some which might not be an issue. Let's just get started with a solid API.
- [x] Get things to compile.
- [x] Improve `document::Value::Number` and `document::Object` types.
- [x] Find a better home for `patch`, `document`, and `selection` modules.
- [x] Better `ReturnSet` name.
- [x] Better way to model `primary_key` and `foreign_keys`.
- [x] Use pointers instead of properties.
- [x] Find a better range implementation.

### CODE REVIEW

File: src/request.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    /// A filter limiting the number of documents deleted.
+    filter: Filter,
+
+    /// What properties to be returned of the deleted documents.
+    returning: ReturnSet
+  }
+}
+
+/// Defines what the request expects to be returned.
+pub enum ReturnSet {
```
Comment (calebmer) [inline src/request.rs]: Can you think of a better name? For both `ReturnSet` and `ReturnSet::Some`.

**Comment by calebmer** [comment]
Some ideas (kinda a note to self) 
- Instead of representing foreign relations in `Collection` we should another enumeration to `Value` maybe `Relationship`. The problem with this is it can't be statically represented. This shouldn't matter as our read requests will undergo join decomposition.
- `Property` assumes a flat data structure, it should be swapped out for a `Pointer` where possible.

**Comment by calebmer** [comment]
I actually think using `Box<Filter>` for recursion would be fine. Do you have any objections?

**Comment by calebmer** [comment]
@svmnotn Made some changes, tell me what you think. I didn't check but I _think_ it compiles because I got no lint errors.

Also, what do you know about ranges in rust? I think we either need to move the `Range` struct to its own module or use another more idiomatic implementation.

File: src/driver.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    // TODO: default implementation which verifies that the request is only
+    // affecting a single thing. This may be done by checking if the collection
+    // primary key is specified in a filter.
+    unimplemented!()
+  }
+
+  /// Send multiple requests at once. If one request fails, **all other
+  /// requests must also fail**. If a driver author wishes they may also
+  /// optimize these requests.
+  fn requests(requests: Vec<Request>) -> Result<Vec<Vec<Value>>, Error> {
```
Comment (calebmer) [inline src/driver.rs]: This isn't actually useful. `request_one` is though.

File: src/values.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  /// True or false.
+  Boolean(bool),
+  /// A numeric value, float, integer, whatever.
+  Number(f64),
+  /// A list of characters.
+  String(String),
+  /// A list of values.
+  Array(Vec<Value>),
+  /// A map of key/value pairs.
+  Object(BTreeMap<String, Value>),
```
Comment (calebmer) [inline src/values.rs]: `BTreeMap` probably isn't performant enough for our case. We should try `Vec<(String, Value)>`

File: src/values.rs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  /// Remove a property from the database entirely.
+  Remove(Pointer)
+}
+
+/// A recursive filter condition for a `Value`.
+pub enum Filter {
+  /// Combine multiple filters with an “and” operator.
+  And(Vec<Box<Filter>>),
+  /// Combine multiple filters with an “or” operator.
+  Or(Vec<Box<Filter>>),
```
Comment (calebmer) [inline src/values.rs]: It shouldn't need `Box` within a `Vec` I think?

**Comment by svmnotn** [comment]
It does compile (Added travis so that we definitely know that)
and you might want to look at [Range](https://doc.rust-lang.org/std/ops/struct.Range.html) and [ops](https://doc.rust-lang.org/std/ops/index.html)

File: src/driver.rs (hunk 2 above)
Comment (svmnotn) [inline src/driver.rs]: so remove it?

Comment (calebmer) [inline src/driver.rs]: Yeah, think so.

**Comment by calebmer** [comment]
@svmnotn that's actually perfect, except is there an enum for all the range-ish types? `Range`, `RangeFrom`, `RangeTo`, and `RangeFull`? Our range can be any of these.

**Comment by svmnotn** [comment]
maybe a trait they all impl?  like [RangeArgument](https://doc.rust-lang.org/collections/range/trait.RangeArgument.html)

**Comment by calebmer** [comment]
Thanks. I'm going to convert the `Request` enum a to driver functions and I'll use that.

**Comment by calebmer** [comment]
Wait, actually `RangeArgument` is unstable…does this matter?

**Comment by svmnotn** [comment]
can you get it to compile? as currently it does not

**Comment by calebmer** [comment]
Yay, it compiles. Think it's ready to merge 🙌 🎉

**Pull request merged**